src/Entity/SmsMessage.php

```diff
@@ -0,0 +1,411 @@
+    $this->recipient->filter(function ($item) use ($recipients) {
+      return !in_array($item->value, $recipients);
+    });
+    return $this;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getOptions() {
+    return ($first = $this->get('options')->first()) ? $first->getValue() : [];
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getOption($name) {
+    $options = $this->getOptions();
+    return isset($options[$name]) ? $options[$name] : NULL;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function setOption($name, $value) {
+    $options = $this->getOptions();
+    $options[$name] = $value;
+    $this->set('options', $options);
+    return $this;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function removeOption($name) {
+    $options = $this->getOptions();
+    unset($options[$name]);
+    $this->set('options', $options);
+    return $this;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getSender() {
+    return ($sender_entity = $this->getSenderEntity()) ? $sender_entity->label() : '';
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function setSender($sender) {
+    // Send to the ether.
+    return $this;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getMessage() {
+    return $this->get('message')->value;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function setMessage($message) {
+    $this->set('message', $message);
+    return $this;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getUuid() {
+    return $this->get('uuid')->value;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getUid() {
+    $sender = $this->getSenderEntity();
+    return ($sender instanceof UserInterface) ? $sender->id() : NULL;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function setUid($uid) {
+    $this->setSenderEntity(User::load($uid));
+    return $this;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function isAutomated() {
+    return $this->get('automated')->value;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function setAutomated($automated) {
+    $this->set('automated', $automated);
+    return $this;
+  }
+
+  /**
+   * Following are implementors of entity interface.
+   *
+   * @see \Drupal\sms\Entity\SmsMessageInterface
+   */
+
+  public function getDirection() {
+    return $this->get('direction');
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getGateway() {
+    return $this->get('gateway');
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function setGateway($gateway) {
+    $this->set('gateway', $gateway);
+    return $this;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getSenderNumber() {
+    return $this->get('sender')->value;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function setSenderNumber($number) {
+    $this->set('sender', $number);
+    return $this;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getSenderEntity() {
```

Comment:
because we arnt necessarily sending a SMS to an entity, the API can send to anyone. Therefore the sender_entity field can be NULL (default value).

Its easier to rely on the 'sender' field being the one and only phone number to send to, when a sms_message record is being created to send to an entity, the api loads the entity and gets its phone number, then saves the sms_message record. this also means we dont have to load the recipient entity when the sms message is being transmitted to the plugin, resulting in lower memory usage.